rts/motoko-rts/src/gc/mark_compact/bitmap.rs

```diff
@@ -85,41 +158,42 @@ pub unsafe fn iter_bits() -> BitmapIter {
 //
 // (We actually need less bits than that as when the heap is full we can't allocate bitmap and mark
 // stack and can't do GC)
-pub const BITMAP_ITER_END: u32 = 1024 * 1024 * 1024;
+pub const BITMAP_ITER_END: u32 = u32::MAX;
 
 impl BitmapIter {
     /// Returns the next bit, or `BITMAP_ITER_END` if there are no more bits set.
     pub fn next(&mut self) -> u32 {
-        debug_assert!(self.current_word_idx <= self.size);
+        debug_assert!(self.current_bit_idx <= self.size);
+
+        if self.current_bit_idx == self.size {
+            return BITMAP_ITER_END;
+        }
 
         // Outer loop iterates 64-bit words
         loop {
-            if self.current_word == 0 && self.current_word_idx == self.size {
-                return BITMAP_ITER_END;
-            }
-
             // Inner loop iterates bits in the current word
             while self.current_word != 0 {
-                if self.current_word & 0b1 == 0b1 {
-                    let bit_idx = (self.current_word_idx * 64) + (64 - self.bits_left);
+                if self.current_word & 0b1 != 0 {
+                    let bit_idx = self.current_bit_idx;
                     self.current_word >>= 1;
-                    self.bits_left -= 1;
+                    self.current_bit_idx += 1;
                     return bit_idx;
                 } else {
                     let shift_amt = self.current_word.trailing_zeros();
                     self.current_word >>= shift_amt;
-                    self.bits_left -= shift_amt;
+                    self.current_bit_idx += shift_amt;
                 }
             }
 
             // Move on to next word
-            self.current_word_idx += 1;
-            if self.current_word_idx == self.size {
+            self.current_bit_idx += self.leading_zeros;
+            if self.current_bit_idx == self.size {
                 return BITMAP_ITER_END;
             }
-            self.current_word =
-                unsafe { *(BITMAP_PTR as *const u64).add(self.current_word_idx as usize) };
-            self.bits_left = 64;
+            self.current_word = unsafe {
```

Comment:
Would it be slightly faster to compute the field values in locals and refer to them via those rather than reading from memory, e.g. for computing self.current_word from the local (hopefully still in a register) rather than self.current_bit_idx? Same for computing self.leading_zeroes  from self.current_word.

Reply:
Below is the function in full, I marked the area you are referring to with comments
``` wasm
  (func $motoko_rts::gc::mark_compact::bitmap::BitmapIter::next::hd19ec51ccf587428 (type 8) (param i32) (result i32)
    (local i32 i32 i32 i64 i64)
    block  ;; label = @1
      local.get 0
      i32.load offset=12
      local.tee 1
      local.get 0
      i32.load offset=8
      local.tee 2
      i32.eq
      br_if 0 (;@1;)
      i32.const 67696
      i32.const 15036
      i32.add
      i32.load
      local.set 3
      local.get 0
      i64.load
      local.set 4
      loop  ;; label = @2
        block  ;; label = @3
          local.get 4
          i64.eqz
          br_if 0 (;@3;)
          block  ;; label = @4
            loop  ;; label = @5
              local.get 4
              i32.wrap_i64
              i32.const 1
              i32.and
              br_if 1 (;@4;)
              local.get 0
              local.get 4
              local.get 4
              i64.ctz
              local.tee 5
              i64.shr_u
              local.tee 4
              i64.store
              local.get 0
              local.get 1
              local.get 5
              i32.wrap_i64
              i32.add
              local.tee 1
              i32.store offset=12
              local.get 4
              i64.eqz
              br_if 2 (;@3;)
              br 0 (;@5;)
            end
          end
          local.get 0
          local.get 1
          i32.const 1
          i32.add
          i32.store offset=12
          local.get 0
          local.get 4
          i64.const 1
          i64.shr_u
          i64.store
          local.get 1
          return
        end
        local.get 0
        local.get 1
        local.get 0
        i32.load offset=16
        i32.add
        local.tee 1
        i32.store offset=12
        local.get 1
        local.get 2
        i32.eq
        br_if 1 (;@1;)
        local.get 0 ;; self
        local.get 3 ;; BITMAP_FORBIDDEN_PTR
        local.get 1 ;; self.current_bit_idx
        i32.const 3
        i32.shr_u ;; / 8
        i32.add ;; .add
        i64.load ;; *
        local.tee 4
        i64.store ;; .current_word =
        local.get 0 ;; self
        local.get 4 ;; self.current_word
        i64.clz;; .leading_zeros()
        i64.store32 offset=16 ;; .leading_zeros =
        br 0 (;@2;)
      end
    end
    i32.const -1)
```
As you can see, the compiler already places all fields in locals:
- `self`: 0
- `current_bit_idx`: 1
- `size`: 2
- `current_word`: 4
- `BITMAP_FORBIDDEN_PTR`: 3

Reply:
Yeah, LLVM does a good job here.